From 15bdf432bc28520cfd7cf2d635c41b7fed303035 Mon Sep 17 00:00:00 2001
From: Joseph Myers <joseph@codesourcery.com>
Date: Tue, 9 Aug 2016 14:33:07 -0700
Subject: [PATCH 03/52] Install extra files for use of mklibs.

EGLIBC patch: <http://www.eglibc.org/archives/patches/msg00431.html>.
Not considered appropriate for upstream glibc.

	Merge from SCB glibc 2.26:

        2015-02-11  James Lemke  <jwlemke@codesourcery.com>

        * Makefiles ($(addprefix
        $(inst_libdir)/libc_pic/,$(install-extras))): Use $(elf-objpfx)
        not $(elfobjdir) in dependency.

        Merge from EGLIBC:

        2008-02-07  Joseph Myers  <joseph@codesourcery.com>

        * Makerules (install-extras, install-map): New variables.
        (installed-libcs): Add libc_pic.a.
        (install-lib): Include _pic.a files for versioned shared
        libraries.
        (install-map-nosubdir, install-extras-nosubdir): Add rules for
        installing extra files.
        (install-no-libc.a-nosubdir): Depend on install-map-nosubdir and
        install-extras-nosubdir.
---
 ChangeLog.csl | 21 +++++++++++++++++++++
 Makerules     | 42 ++++++++++++++++++++++++++++++++++++++++--
 2 files changed, 61 insertions(+), 2 deletions(-)

diff --git a/ChangeLog.csl b/ChangeLog.csl
index a53eef31cc..30eac2e70f 100644
--- a/ChangeLog.csl
+++ b/ChangeLog.csl
@@ -2,6 +2,27 @@
 
 	Merge from SCB glibc 2.26:
 
+	2015-02-11  James Lemke  <jwlemke@codesourcery.com>
+
+        * Makefiles ($(addprefix
+        $(inst_libdir)/libc_pic/,$(install-extras))): Use $(elf-objpfx)
+        not $(elfobjdir) in dependency.
+
+        Merge from EGLIBC:
+
+        2008-02-07  Joseph Myers  <joseph@codesourcery.com>
+
+        * Makerules (install-extras, install-map): New variables.
+        (installed-libcs): Add libc_pic.a.
+        (install-lib): Include _pic.a files for versioned shared
+        libraries.
+        (install-map-nosubdir, install-extras-nosubdir): Add rules for
+        installing extra files.
+        (install-no-libc.a-nosubdir): Depend on install-map-nosubdir and
+        install-extras-nosubdir.
+
+	Merge from SCB glibc 2.26:
+
 	2015-03-23  Sandra Loosemore  <sandra@codesourcery.com>
 
         * elf/rtld.c (_dl_start): Zero-initialize the entire bootstrap_map
diff --git a/Makerules b/Makerules
index ef6abeac6d..a8ccbf81fe 100644
--- a/Makerules
+++ b/Makerules
@@ -778,6 +778,9 @@ ifeq ($(build-shared),yes)
 $(common-objpfx)libc.so: $(common-objpfx)libc.map
 endif
 common-generated += libc.so libc_pic.os
+ifndef subdir
+install-extras := soinit.o sofini.o
+endif
 ifdef libc.so-version
 $(common-objpfx)libc.so$(libc.so-version): $(common-objpfx)libc.so
 	$(make-link)
@@ -1009,6 +1012,7 @@ endef
 installed-libcs := $(foreach o,$(filter-out .os,$(object-suffixes-for-libc)),\
 			     $(inst_libdir)/$(patsubst %,$(libtype$o),\
 						     $(libprefix)$(libc-name)))
+installed-libcs := $(installed-libcs) $(inst_libdir)/libc_pic.a
 
 .PHONY: check-install-supported
 check-install-supported:
@@ -1057,6 +1061,22 @@ versioned := $(strip $(foreach so,$(install-lib.so),\
 install-lib.so-versioned := $(filter $(versioned), $(install-lib.so))
 install-lib.so-unversioned := $(filter-out $(versioned), $(install-lib.so))
 
+# Install the _pic.a files for versioned libraries, and corresponding
+# .map files.
+# libpthread_pic.a breaks mklibs, so don't install it and its map.
+install-lib := $(install-lib) $(install-lib.so-versioned:%.so=%_pic.a)
+install-lib := $(filter-out libpthread_pic.a,$(install-lib))
+# Despite having a soname libhurduser and libmachuser do not use symbol
+# versioning, so don't install the corresponding .map files.
+ifeq ($(build-shared),yes)
+install-map := $(patsubst %.so,%.map,\
+			$(foreach L,$(install-lib.so-versioned),$(notdir $L)))
+install-map := $(filter-out libhurduser.map libmachuser.map libpthread.map,$(install-map))
+ifndef subdir
+install-map := $(install-map) libc.map
+endif
+endif
+
 # For versioned libraries, we install three files:
 #	$(inst_libdir)/libfoo.so	-- for linking, symlink or ld script
 #	$(inst_slibdir)/libfoo.so.NN	-- for loading by SONAME, symlink
@@ -1311,9 +1331,22 @@ $(addprefix $(inst_includedir)/,$(headers-nonh)): $(inst_includedir)/%: \
 endif	# headers-nonh
 endif	# headers
 
+ifdef install-map
+$(addprefix $(inst_libdir)/,$(patsubst lib%.map,lib%_pic.map,$(install-map))): \
+  $(inst_libdir)/lib%_pic.map: $(common-objpfx)lib%.map $(+force)
+	$(do-install)
+endif
+
+ifdef install-extras
+$(addprefix $(inst_libdir)/libc_pic/,$(install-extras)): \
+  $(inst_libdir)/libc_pic/%.o: $(elf-objpfx)%.os $(+force)
+	$(do-install)
+endif
+
 .PHONY: install-bin-nosubdir install-bin-script-nosubdir \
 	install-rootsbin-nosubdir install-sbin-nosubdir install-lib-nosubdir \
-	install-data-nosubdir install-headers-nosubdir
+	install-data-nosubdir install-headers-nosubdir install-map-nosubdir \
+	install-extras-nosubdir
 install-bin-nosubdir: $(addprefix $(inst_bindir)/,$(install-bin))
 install-bin-script-nosubdir: $(addprefix $(inst_bindir)/,$(install-bin-script))
 install-rootsbin-nosubdir: \
@@ -1326,6 +1359,10 @@ install-data-nosubdir: $(addprefix $(inst_datadir)/,$(install-data))
 install-headers-nosubdir: $(addprefix $(inst_includedir)/,$(headers))
 install-others-nosubdir: $(install-others)
 install-others-programs-nosubdir: $(install-others-programs)
+install-map-nosubdir: $(addprefix $(inst_libdir)/,\
+		       $(patsubst lib%.map,lib%_pic.map,$(install-map)))
+install-extras-nosubdir: $(addprefix $(inst_libdir)/libc_pic/,\
+		       $(install-extras))
 
 # We need all the `-nosubdir' targets so that `install' in the parent
 # doesn't depend on several things which each iterate over the subdirs.
@@ -1335,7 +1372,8 @@ install-%:: install-%-nosubdir ;
 
 .PHONY: install install-no-libc.a-nosubdir
 install-no-libc.a-nosubdir: install-headers-nosubdir install-data-nosubdir \
-			    install-lib-nosubdir install-others-nosubdir
+			    install-lib-nosubdir install-others-nosubdir \
+			    install-map-nosubdir install-extras-nosubdir
 ifeq ($(build-programs),yes)
 install-no-libc.a-nosubdir: install-bin-nosubdir install-bin-script-nosubdir \
 			    install-rootsbin-nosubdir install-sbin-nosubdir \
-- 
2.11.1

